pyro/contrib/oed/eig.py

```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
+        else:
+            terms -= sum(cond_trace.nodes[l]["log_prob"] for l in observation_labels)
+
+        return safe_mean_terms(terms)
+
+    return loss_fn
+
+
+def lfire_loss(model_marginal, model_conditional, h, observation_labels, target_labels):
+
+    def loss_fn(design, num_particles, evaluation=False, **kwargs):
+
+        try:
+            pyro.module("h", h)
+        except AssertionError:
+            pass
+
+        expanded_design = lexpand(design, num_particles)
+        model_conditional_trace = poutine.trace(model_conditional).get_trace(expanded_design)
+
+        if not evaluation:
+            model_marginal_trace = poutine.trace(model_marginal).get_trace(expanded_design)
+
+            h_joint = h(expanded_design, model_conditional_trace, observation_labels, target_labels)
+            h_independent = h(expanded_design, model_marginal_trace, observation_labels, target_labels)
+
+            terms = torch.nn.functional.softplus(-h_joint) + torch.nn.functional.softplus(h_independent)
+            return safe_mean_terms(terms)
+
+        else:
+            h_joint = h(expanded_design, model_conditional_trace, observation_labels, target_labels)
+            return safe_mean_terms(h_joint)
+
+    return loss_fn
+
+
+def vnmc_eig_loss(model, guide, observation_labels, target_labels):
+
+    def loss_fn(design, num_particles, evaluation=False, **kwargs):
+        N, M = num_particles
+        expanded_design = lexpand(design, N)
+
+        # Sample from p(y, theta | d)
+        trace = poutine.trace(model).get_trace(expanded_design)
+        y_dict = {l: lexpand(trace.nodes[l]["value"], M) for l in observation_labels}
+
+        # Sample M times from q(theta | y, d) for each y
+        reexpanded_design = lexpand(expanded_design, M)
+        conditional_guide = pyro.condition(guide, data=y_dict)
+        guide_trace = poutine.trace(conditional_guide).get_trace(
+            y_dict, reexpanded_design, observation_labels, target_labels)
+        theta_y_dict = {l: guide_trace.nodes[l]["value"] for l in target_labels}
+        theta_y_dict.update(y_dict)
+        guide_trace.compute_log_prob()
+
+        # Re-run that through the model to compute the joint
+        modelp = pyro.condition(model, data=theta_y_dict)
+        model_trace = poutine.trace(modelp).get_trace(reexpanded_design)
+        model_trace.compute_log_prob()
+
+        terms = -sum(guide_trace.nodes[l]["log_prob"] for l in target_labels)
+        terms += sum(model_trace.nodes[l]["log_prob"] for l in target_labels)
+        terms += sum(model_trace.nodes[l]["log_prob"] for l in observation_labels)
+        terms = -terms.logsumexp(0) + math.log(M)
+
+        # At eval time, add p(y | theta, d) terms
+        if evaluation:
+            trace.compute_log_prob()
+            terms += sum(trace.nodes[l]["log_prob"] for l in observation_labels)
+
+        return safe_mean_terms(terms)
+
+    return loss_fn
+
+
+def safe_mean_terms(terms):
```

Comment:
nit: if this isn't used elsewhere and isn't meant to be called by users, rename to `_safe_mean_terms`